rust/lance-index/src/scalar/inverted.rs

```diff
@@ -361,91 +378,82 @@ impl InvertedList {
             Arc::new(schema),
             vec![
                 Arc::new(token_id_col) as ArrayRef,
-                Arc::new(row_ids_col.finish()) as ArrayRef,
-                Arc::new(frequencies_col.finish()) as ArrayRef,
+                Arc::new(row_ids_col) as ArrayRef,
+                Arc::new(frequencies_col) as ArrayRef,
             ],
         )?;
         Ok(batch)
     }
 
     async fn load(reader: Arc<dyn IndexReader>) -> Result<Self> {
-        let mut tokens = Vec::new();
-        let mut row_ids_list = Vec::new();
-        let mut frequencies_list = Vec::new();
+        let mut inverted_list = HashMap::new();
         for i in 0..reader.num_batches().await {
             let batch = reader.read_record_batch(i).await?;
             let token_col = batch[TOKEN_ID_COL].as_primitive::<datatypes::UInt32Type>();
             let row_ids_col = batch[ROW_ID].as_list::<i32>();
             let frequencies_col = batch[FREQUENCY_COL].as_list::<i32>();
 
-            tokens.extend(token_col.iter().map(|v| v.unwrap()));
-            for value in row_ids_col.iter() {
-                let value = value.unwrap();
-                let row_ids = value
-                    .as_primitive::<datatypes::UInt64Type>()
-                    .values()
-                    .iter()
-                    .cloned()
-                    .collect_vec();
-                row_ids_list.push(row_ids);
-            }
-            for value in frequencies_col.iter() {
-                let value = value.unwrap();
-                let frequencies = value
-                    .as_primitive::<datatypes::UInt64Type>()
-                    .values()
+            for ((token_id, row_ids), frequencies) in token_col
+                .iter()
+                .zip(row_ids_col.iter())
+                .zip(frequencies_col.iter())
+            {
+                let token_id = token_id.unwrap();
```

Comment:
Same thing here, if it's non-null:
```suggestion
            for ((token_id, row_ids), frequencies) in token_col
                .values()
                .iter()
                .zip(row_ids_col.iter())
                .zip(frequencies_col.iter())
            {
```